File: godottenjin/src/main/java/com/onecat/godottenjin/GodotTenjin.java
package com.onecat.godottenjin;

import android.app.Activity;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.ads.identifier.AdvertisingIdClient.Info;
import com.google.android.gms.ads.identifier.AdvertisingIdClient;

import com.tenjin.android.Callback;
import com.tenjin.android.TenjinSDK;

import java.lang.Exception;
import java.util.HashSet;
import java.util.Set;
import java.util.Map;

import org.godotengine.godot.Godot;
import org.godotengine.godot.plugin.GodotPlugin;
import org.godotengine.godot.plugin.SignalInfo;
import org.godotengine.godot.plugin.UsedByGodot;

@SuppressWarnings({"unused", "SpellCheckingInspection"})
public class GodotTenjin extends GodotPlugin
{
	private final Activity activity;
    private String apiKey = null;

    public GodotTenjin(Godot godot)
    {
        super(godot);
        activity = getActivity();
    }

    private TenjinSDK getTenjinInstance() {
        return TenjinSDK.getInstance(activity, apiKey);
    }

    @Nullable
    @Override
    public View onMainCreate(Activity activity) {
        return null;
    }

/*    @Override
    public void onMainResume() {

        if (apiKey != null && activity != null) {
            TenjinSDK instance = getTenjinInstance();
            instance.connect();
        }
    }

 */

    @NonNull
    @Override
    public String getPluginName() {
        return "GodotTenjin";
    }

    @NonNull
    @Override
    public Set<SignalInfo> getPluginSignals() {
        Set<SignalInfo> signalInfoSet = new HashSet<>();
        signalInfoSet.add(new SignalInfo("plugin_initialized"));
        signalInfoSet.add(new SignalInfo("gaid_obtain_failed", String.class));
        return  signalInfoSet;
    }

    @UsedByGodot
    public void initialize(final String _apiKey, final String _deepLinkUri, final int appstoreType) {
        apiKey = _apiKey;
        TenjinSDK instance = TenjinSDK.getInstance(activity, apiKey);
        TenjinSDK.AppStoreType _appstoretype = switch (appstoreType) {
            case 1 -> TenjinSDK.AppStoreType.googleplay;
            case 2 -> TenjinSDK.AppStoreType.amazon;
            case 3 -> TenjinSDK.AppStoreType.other;
            default -> TenjinSDK.AppStoreType.unspecified;
        };
        instance.setAppStore(_appstoretype);
        if(_deepLinkUri != null && !_deepLinkUri.isEmpty()) {
            instance.connect(_deepLinkUri);
        } else {
            instance.connect();
        }
        emitSignal("plugin_initialized");
        instance.getDeeplink(new Callback() {
            @Override
            public void onSuccess(boolean clickedTenjinLink, boolean isFirstSession, Map<String, String> data) {
                if (clickedTenjinLink) {
                    if (isFirstSession) {
                        if (data.containsKey(TenjinSDK.DEEPLINK_URL)) {
                           // deferred_deeplink_url logic here
                        }
                    }
                }
            }
        });
    }

    @UsedByGodot
    public void logEvent(final String event) {
        if(apiKey != null)
            getTenjinInstance().eventWithName(event);
    }

    @UsedByGodot
    public void logEventWithValue(final String event, final int value) {
        if(apiKey != null)
            getTenjinInstance().eventWithNameAndValue(event, value);
    }

    @UsedByGodot
    public void logPurchase(final String productId, final String currencyCode, final int quantity, final float unitPrice) {
        if(apiKey != null)
            getTenjinInstance().transaction(productId, currencyCode, quantity, unitPrice);
    }

    @UsedByGodot
    public void logPurchaseWithSignature(final String productId, final String currencyCode, final int quantity, final float unitPrice, final String purchaseData, final String dataSignature) {
        if(apiKey != null)
            getTenjinInstance().transaction(productId, currencyCode, quantity, unitPrice, purchaseData, dataSignature);
    }

    @UsedByGodot
    public String getGAID()
    {
        Info adInfo;
        try {
            adInfo = AdvertisingIdClient.getAdvertisingIdInfo(activity);
            return adInfo.getId();
        } catch (Exception e) {
            emitSignal("gaid_obtain_failed", e.toString());
        }
        return "";
    }

}

File: godottenjin/src/main/kotlin/com/onecat/tenjin/GodotTenjin.kt
package com.onecat.tenjin

import android.app.Activity
import android.util.Log
import android.view.View
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.google.android.gms.ads.identifier.AdvertisingIdClient.Info
import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.tenjin.android.TenjinSDK
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.UsedByGodot
import java.io.IOException

class GodotTenjin(activity:Godot) : GodotPlugin(activity) {
    private var apiKey : String? = null

    override fun getPluginName() : String {
        return "GodotTenjin"
    }

    @UsedByGodot
    fun getAdID() : String? {
        val adInfo : Info?
        try {
            adInfo = activity?.let { AdvertisingIdClient.getAdvertisingIdInfo(it) }
            return adInfo?.id
        } catch (e: IOException) {
            Log.e("godot", e.toString())
        } catch (e: GooglePlayServicesNotAvailableException) {
            Log.e("godot", e.toString())
        } catch (e: Exception) {
            Log.e("godot", e.toString())
        }
        return ""
    }

    @UsedByGodot
    fun initializePlugin(_apiKey : String, _deepLinkUri : String) {
        apiKey = _apiKey
        val sdkInstance : TenjinSDK = TenjinSDK.getInstance(activity, apiKey)
        sdkInstance.setAppStore(TenjinSDK.AppStoreType.googleplay)
        if (_deepLinkUri != "") {
            sdkInstance.connect(_deepLinkUri)
        }
        else {
            sdkInstance.connect()
        }

        Log.i("godot", "Godot Tenjin plugin initialized")
        Log.i("godot", "GAID:"+getAdID())
        sdkInstance.getDeeplink { clickedTenjinLink, isFirstSession, data ->
            if (clickedTenjinLink) {
                if (isFirstSession) {
                    if (data.containsKey(TenjinSDK.DEEPLINK_URL)) {
                        // direct to app part here
                    }
                }
            }
        }
    }

    @UsedByGodot
    fun logEvent(event:String) {
        if (apiKey == "" || apiKey == null){
            Log.e("godot", "Tenjin: attempt to logEvent with no API key, did nothing")
            return
        }
        tenjinInstance().eventWithName(event)
    }

    @UsedByGodot
    fun logEventWithValue(event:String, value:Int) {
        if (apiKey == "" || apiKey == null){
            Log.e("godot", "Tenjin: attempt to logEventWithValue with no API key, did nothing")
            return
        }
        tenjinInstance().eventWithNameAndValue(event, value)
    }

    @UsedByGodot
    fun logPurchase(productId:String, currencyCode:String, quantity:Int, unitPrice:Double) {
        if (apiKey == "" || apiKey == null){
            Log.e("godot", "Tenjin: attempt to logPurchase an event with no API key, did nothing")
            return
        }
        tenjinInstance().transaction(productId, currencyCode, quantity, unitPrice)
    }

    @UsedByGodot
    fun logPurchaseWithSignature(
        productId:String,
        currencyCode:String,
        quantity:Int, unitPrice:Double,
        purchaseData:String,
        dataSignature:String) {
        tenjinInstance().transaction(productId, currencyCode, quantity, unitPrice, purchaseData, dataSignature)
    }

    private fun tenjinInstance() : TenjinSDK {
        return TenjinSDK.getInstance(activity, apiKey)
    }
}
